### ISSUE TITLE
Consider splitting `azurerm_application_gateway` resource into several smaller ones

### ISSUE THREAD

**Issue by maxbog** [hashicorp/terraform-provider-azurerm#14668]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I've been working with Application Gateway lately and wanted to defer some responsibility of configuring parts of Application Gateway to other teams. Right now, this is only possible by making the app gateway resource a giant, shared one, which everyone can change. I'd like to split the app gateway into several sub-resources, similar to what happened to IotHub resource: https://github.com/hashicorp/terraform-provider-azurerm/issues/3303

I'd like to be able to define separate resources for backends, frontends, routing rules, health probes. This would reverse the dependencies between the app gateway and backend services, making the tf code more modular. Dependencies between sub-resources would also be much clearer. Now, they are only ensured by the consistent naming of the sections of app_gateway configuration. Splitting the resource would make those dependencies explicit and easier to get wrong.

Deployments would also be independent in a scenario where a single app gateway is shared by multiple teams. The teams can reconfigure the backends or routing as needed and apply their part of the infrastructure change without affecting other teams.

I am willing to gradually implement the resources, but want to get an OK/not OK from the maintainers.

### New or Affected Resource(s)

* azurerm_application_gateway
* azurerm_application_gateway_backend_address_pool
* azurerm_application_gateway_backend_http_settings
* azurerm_application_gateway_probe
* azurerm_application_http_listener
* azurerm_application_request_routing_rule
* azurerm_application_frontend_port
* azurerm_application_frontend_ip_configuration
* azurerm_application_frontend_ssl_certificate

### Potential Terraform Configuration

Example is adopted from [the one provided in `azurerm_application_gateway` documentation](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/application_gateway#example-usage)

```hcl
resource "azurerm_resource_group" "example" {
  name     = "example-resources"
  location = "West Europe"
}

resource "azurerm_virtual_network" "example" {
  name                = "example-network"
  resource_group_name = azurerm_resource_group.example.name
  location            = azurerm_resource_group.example.location
  address_space       = ["10.254.0.0/16"]
}

resource "azurerm_subnet" "frontend" {
  name                 = "frontend"
  resource_group_name  = azurerm_resource_group.example.name
  virtual_network_name = azurerm_virtual_network.example.name
  address_prefixes     = ["10.254.0.0/24"]
}

resource "azurerm_subnet" "backend" {
  name                 = "backend"
  resource_group_name  = azurerm_resource_group.example.name
  virtual_network_name = azurerm_virtual_network.example.name
  address_prefixes     = ["10.254.2.0/24"]
}

resource "azurerm_public_ip" "example" {
  name                = "example-pip"
  resource_group_name = azurerm_resource_group.example.name
  location            = azurerm_resource_group.example.location
  allocation_method   = "Dynamic"
}

resource "azurerm_application_gateway" "network" {
  name                = "example-appgateway"
  resource_group_name = azurerm_resource_group.example.name
  location            = azurerm_resource_group.example.location

  sku {
    name     = "Standard_Small"
    tier     = "Standard"
    capacity = 2
  }

  gateway_ip_configuration {
    name      = "my-gateway-ip-configuration"
    subnet_id = azurerm_subnet.frontend.id
  }
}

resource "application_gateway_frontend_port" "http" {
    application_gateway_name  = azurerm_application_gateway.network.name
    resource_group_name       = azurerm_application_gateway.network.resource_group_name

    name = "http"
    port = 80
}

resource "application_gateway_frontend_ip_configuration" "feip" {
    application_gateway_name = azurerm_application_gateway.network.name
    resource_group_name      = azurerm_application_gateway.network.resource_group_name

    name                 = "${azurerm_virtual_network.example.name}-feip"
    public_ip_address_id = azurerm_public_ip.example.id
}

resource "application_gateway_http_listener" "httplstn" {
    application_gateway_name = azurerm_application_gateway.network.name
    resource_group_name      = azurerm_application_gateway.network.resource_group_name

    name                           = "${azurerm_virtual_network.example.name}-httplstn"
    frontend_ip_configuration_name = application_gateway_frontend_port.feip.name
    frontend_port_name             = application_gateway_frontend_port.http.name
    protocol                       = "Http"
}

resource "application_gateway_backend_address_pool" "beap" {
    application_gateway_name = azurerm_application_gateway.network.name
    resource_group_name      = azurerm_application_gateway.network.resource_group_name

    name = "${azurerm_virtual_network.example.name}-beap"
}

resource "application_gateway_backend_http_settings" "be-htst" {
    application_gateway_name = azurerm_application_gateway.network.name
    resource_group_name      = azurerm_application_gateway.network.resource_group_name

    name                  = "${azurerm_virtual_network.example.name}-be-htst"
    cookie_based_affinity = "Disabled"
    path                  = "/path1/"
    port                  = 80
    protocol              = "Http"
    request_timeout       = 60
}

resource "application_gateway_request_routing_rule" "rqrt" {
    application_gateway_name = azurerm_application_gateway.network.name
    resource_group_name      = azurerm_application_gateway.network.resource_group_name

    name                       = "${azurerm_virtual_network.example.name}-rqrt"
    rule_type                  = "Basic"
    http_listener_name         = application_gateway_http_listener.httplstn.name
    backend_address_pool_name  = application_gateway_backend_address_pool.beap.name
    backend_http_settings_name = application_gateway_backend_http_settings.be-htst.name
}
```

### References

* #3303

**Comment by tombuildsstuff** [comment]
hi @maxbog 

Thanks for opening this issue.

When provisioning an Application Gateway unfortunately it's required to create at least one of most of the sub-components, as such it's not possible to add support for these as separate resources without introducing a perectual diff on the main `azurerm_application_gateway` resource. This has come up a number of times over the years and is unfortunately blocked on the Azure Service Team who need to add support for this - which is currently being tracked [in this upstream issue](https://github.com/Azure/azure-rest-api-specs/issues/2313).

Unfortunately the Azure Networking team have since tried to move this Feature Request to the Azure Feedback site, which has since been shut down - as such [I've just asked for an update](https://github.com/Azure/azure-rest-api-specs/issues/2313#issuecomment-997844834) to get this re-opened - however it may also be worth opening a new issue on that repository too.

Since the API doesn't allow provisioning an Application Gateway without any subcomponents at the moment unfortunately and so is a feature request for the Azure Service Team - I'm going to close this issue for the moment, but I'd recommend either commenting on the existing issue above, or making a new one so that the Service Team can take a look - but unfortunately there's not much we can do from Terraform's side at the moment.

Thanks!
